### PR TITLE
[Clang][Sema] Abbreviated function templates do not append invented parameters to empty template parameter lists

### PR DESCRIPTION
**Pull request by sdkrystian**
According to [[dcl.fct] p23](http://eel.is/c++draft/dcl.fct#23):
> An abbreviated function template can have a _template-head_. The invented _template-parameters_ are appended to the _template-parameter-list_ after the explicitly declared _template-parameters_.

`template<>` is not a _template-head_ -- a _template-head_ must have at least one _template-parameter_. This patch corrects our current behavior of appending the invented template parameters to the innermost template parameter list, regardless of whether it is empty. Example:
```cpp
template<typename T>
struct A 
{ 
    void f(auto); 
};

template<>
void A<int>::f(auto); // ok

template<>
template<> // warning: extraneous template parameter list in template specialization
void A<int>::f(auto);
```

### CODE REVIEW

File: clang/lib/Sema/SemaDecl.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaDecl.cpp:9762]:
```suggestion
    if (!TemplateParamLists.empty() && !TemplateParamLists.back()->empty() &&
```

File: clang/lib/Sema/SemaDeclCXX.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaDeclCXX.cpp:19306]:
```suggestion
  if (ExplicitParams && !ExplicitParams->empty()) {
```

File: clang/lib/Sema/SemaDecl.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaDecl.cpp:9762]: Ah, urgh, i hate 'sorta vector' types.  Yeah, i think 'empty' makes a lot of sense, it improves the readability a ton in a situation like this.

**Review by erichkeane — APPROVED** [review]

File: clang/lib/Sema/SemaDecl.cpp
Comment (sdkrystian) [reply clang/lib/Sema/SemaDecl.cpp]: Added. I'll open a separate PR that converts other uses of `size()` to use `empty()` where appropriate

**Review by erichkeane — APPROVED** [review]

**Comment by sdkrystian** [comment]
Looks like there is a test (`clang-tools-extra/test/clang-tidy/infrastructure/diagnostic.cpp`) that contains the following:
```cpp
#ifdef PR64602 // Should not crash
template <class T = void>
struct S
{
    auto foo(auto);
};

template <>
auto S<>::foo(auto)
{
    return 1;
}
// CHECK8: error: template parameter list matching the non-templated nested type 'S<>' should be empty ('template<>') [clang-diagnostic-error]
#endif
```
Which is failing because we now (correctly) accept this. Can I open a PR that addresses this single test, or would I need to revert and reapply the patch @erichkeane ?

**Comment by erichkeane** [comment]
> Looks like there is a test (`clang-tools-extra/test/clang-tidy/infrastructure/diagnostic.cpp`) that contains the following:
> 
> ```c++
> #ifdef PR64602 // Should not crash
> template <class T = void>
> struct S
> {
>     auto foo(auto);
> };
> 
> template <>
> auto S<>::foo(auto)
> {
>     return 1;
> }
> // CHECK8: error: template parameter list matching the non-templated nested type 'S<>' should be empty ('template<>') [clang-diagnostic-error]
> #endif
> ```
> 
> Which is failing because we now (correctly) accept this. Can I open a PR that addresses this single test, or would I need to revert and reapply the patch @erichkeane ?

If you act quickly, you can just do a new PR and get quick review from the code owners on clang-tidy, which would not require a revert.

**Comment by nico** [comment]
This breaks check-clang-tools: http://45.33.8.238/linux/130320/step_8.txt

Please take a look and revert for now if it takes a while to fix.

**Pull request merged**
